File: 20.rb
def is_valid(s)
    return true if s.length < 1
    return false if s.length == 1
    array = s.split("")
    stack = [array.shift]
    until array.length < 1
        case array.first
        when "("
            stack.push(array.first)
            array.shift
        when "{"
            stack.push(array.first)
            array.shift
        when "["
            stack.push(array.first)
            array.shift
        when ")"
            return false if stack.last != "("
            stack.pop
            array.shift
        when "}"
            return false if stack.last != "{"
            stack.pop
            array.shift
        when "]"
            return false if stack.last != "["
            stack.pop
            array.shift
        end
    end
    return true if stack.empty?
    return false
end

File: Easy/20.js
var isValid = function(s) {
    let h = {};
    h["("] = 0
    h["{"] = 0
    h["["] = 0
    for (let i = 0; i < s.length; i++) {
        let char = s[i];
        switch (char) {
            case "(":
                h["("]++;
                break;
            case "{":
                h["{"]++;
                break;
            case "[":
                h["["]++;
                break;
            case ")":
                if (h["("] < 1) {return false}
                h["("]--;
                break;
            case "}":
                if (h["}"] < 1) {return false}
                h["{"]--;
                break;
            case "]":
                if (h["]"] < 1) {return false}
                h["["]--;
                break;
        }
    }
    return Object.values(h).every(val => val === 0)
};

console.log(isValid("(){}[]"))